share/picongpu/tests/Synchrotron_plugin/lib/validate.py

```diff
@@ -0,0 +1,133 @@
+    return x0, np.abs(y0 - y1) / y0
+
+
+# read the parameters from the params.txt file: gamma, Heff, dt
+def read_setup(folderName):
+    # in params.txt we have the following parameters:
+    # gamma: 1000, Heff: 1000000000000000.0, dt: 4.830524160012079e-17
+
+    with open(folderName + "/params.txt") as f:
+        lines = f.readlines()
+        data = lines[0].split(",")
+        gamma = data[0].split(":")[1].strip()
+        Heff = data[1].split(":")[1].strip()
+        dt = data[2].split(":")[1].strip()
+
+    return float(gamma), float(Heff), float(dt)
+
+
+iterNo = 4000  # iteration from which we read the data must be the same as in ../bin/validate.ci
+
+
+def main(dataPath):
+    print(f"Validating data in {dataPath}")
+    gamma, Heff, dt = read_setup(dataPath)
+
+    series = io.Series(dataPath + "/openPMD/simData_test_%T.bp", io.Access.read_only)
+    it = series.iterations[iterNo]
+
+    simulation_dt = it.get_attribute("dt") * it.get_attribute("unit_time")
+    if abs(simulation_dt - dt) > 1e-10:  # check if the simulation dt is the same as the expected dt. 1e-10 is arbitrary
+        raise SystemExit(f"Simulation dt is {simulation_dt} but expected dt is {dt}\n Test failed")
+
+    h = it.particles["y"]  # photons
+    e = it.particles["e"]  # electrons
+
+    # photons
+    h1 = h["weighting"][io.Mesh_Record_Component.SCALAR]
+    w = h1[:]
+    w_SI = h1.unit_SI
+    series.flush()
+    w *= w_SI
+    h1 = h["momentum"]["y"]
+    p_y = h1[:]
+    p_y_SI = h1.unit_SI
+    series.flush()
+    p_y = np.float64(p_y)
+    p_y *= p_y_SI / w
+    hist_data = np.abs(p_y)
+    if len(hist_data) < 5e5:
+        raise SystemExit(
+            f"Number of photons is {len(hist_data)} but expected number of photons to be at least 5e5\n Test failed"
+        )
+    hist_data = np.abs(p_y * const.c / const.elementary_charge)  # calculate the energy of photons
+
+    # electron energy: take electron momentum and calculate energy
+    h1 = e["weighting"][io.Mesh_Record_Component.SCALAR]
+    e_w = h1[:]  # get the weighting
+    e_w_SI = h1.unit_SI  # get the weighting unit
+    series.flush()
+    e_w *= e_w_SI  # convert to SI units
+    h1 = e["momentum"]["y"]
+    e_p_y = h1[:]  # get the momentum
+    e_p_y_SI = h1.unit_SI  # get the momentum unit
+    series.flush()
+    e_p_y *= e_p_y_SI / e_w
+    Energy_e = momentum_to_energy(e_p_y, const.m_e).max()  # convert momentum to energy
+
+    hist_data /= Energy_e  # go to delta space
+
+    min_exp = np.floor(np.log10(np.min(hist_data)))
+    max_exp = np.ceil(np.log10(np.max(hist_data)))
+    bins = np.logspace(min_exp, max_exp, 100)
+    a, b = np.histogram(hist_data, bins=bins)
+    normaliztion_factor = iterNo * len(e_w)
+
+    delta = bins
+    analytical_integrated = []
+    for x0, x1 in zip(delta[:-1], delta[1:]):
+        analytical_integrated.append(quad(lambda x: analytical_Propability(x, gamma, Heff, dt), x0, x1)[0])
```

Comment:
```suggestion
    analytical_integrated = [quad(lambda x: analytical_Propability(x, gamma, Heff, dt), x0, x1)[0]) for x0, x1 in zip(delta[:-1], delta[1:])]
```

Reply:
What is the advantage of this code suggestions @chillenzer except that it is shorter?

Reply:
Typically, e.g. [here](https://scribe.rip/how-list-comprehensions-can-help-your-code-look-better-and-run-smoother-3cf8f87172ae), 3 (and a half) reasons are mentioned for this kind of refactoring:
1. It's shorter as you pointed out already.
2. It's faster due to pre-allocation (tiny bit) and avoiding repeated calls to `.append()` (majority).
3. It's easier to read and helps to avoid mixing the independent creation of a list with unrelated functionality just because they happen to have the same underlying for-loop (which would be a maintenance burden in the future).

The last item "easier to read" isn't always true which makes up for the half reason: If it is not true, it is likely because the functionality inside of the for-loop was too verbose to be easily comprehensible (or even fit) if written in one line. In that case, the function most likely violates SRP (with the implications stated elsewhere in this PR already). So, the half reason is that it points you to places where one should extract functionality into a separate function.

Coming to think about it, the `quad` call could be such an extraction candidate.